### ISSUE TITLE
Back up databases to S3-compatible storage

### ISSUE THREAD

**Issue by marcua** [marcua/ayb#14]
There are many options for backing up SQLite to S3. The one that most directly focuses on continuous streaming backups and restore (as opposed to maintaining read replicas or multiple writers, etc.) is [litestream](https://litestream.io/getting-started/), but there are others like [libSQL's bottomless replication work-in-progress](https://github.com/tursodatabase/libsql/tree/main/bottomless).

We discuss alternatives and why we aren't (yet) considering them below, but for now, we will go with [simple scheduled backups](https://litestream.io/alternatives/cron/).
* In `ayb.toml`, you will configure: S3-compatible storage details, frequency of backup, and how many backups to keep before pruning old ones.
* `ayb` will have a Snapshot data model that can store details on database snapshots (when we took it, what the hash of the database directory was, whether it was a scheduled or manual snapshot, the path to the backup on S3-compatible storage, etc.)
* On the `ayb` server, we'll run a periodic task to check if a database has changed recently (based on the hash of its database directory, but perhaps there's a way to query for the most recent transaction ID instead) and if it has, create a new snapshot.
* The backup in SQLite will happen via [VACUUM](https://www.sqlite.org/lang_vacuum.html), which will reduce file size and won't potentially run forever under high write loads like the [online backup API](https://www.sqlite.org/backup.html).
* The advantages of this approach are:
  * Simplest implementation path.
  * No dependence on other tools that are being rapidly developed and don't have documented APIs (yet).
  * Due to the simplicity and generality of the approach, it shouldn't be hard to apply it to other database types (e.g., DuckDB) that have less developed stream-the-WAL-to-S3 functionality.
  * I suspect that in practice, most databases will be write-dormant most of the time given the target use cases, so this might require less overhead than tools that constantly monitor a database log for changes.
* The disadvantages of the approach are:
  * Potentially more work per backup: if you only update a row, we still VACUUM and export the entire database.
  * Periodic snapshots have a larger window for data loss on the order of tens of minutes to an hour, whereas shipping the write-ahead log typically results in only seconds of data loss.
  * Related to the previous reason, implementing things like multi-node `ayb` support with automatic failover will be difficult in the future in cases where a node becomes inaccessible and you're not sure how many minutes/tens of minutes of backups are unsynced.

The plan
- [x] Ask on SQLite forum about how to find the most recent txid/modification: https://sqlite.org/forum/forumpost/2798df4be8. The approach to beat is to have each DB in its own directory (SQLite spans multiple files), and use filesystem modified times / hashes of all files in the directory to determine changes.
- [x] Refactors to directory structure that support snapshots and SQLite-in-directory (https://github.com/marcua/ayb/pull/311)
- [x] Refactors to allow dangerous/un-isolated requests to complete the backups (requires ATTACH) (https://github.com/marcua/ayb/pull/312)
- [x] Backups to the local filesystem (https://github.com/marcua/ayb/pull/316)
  - [x] `periodic_snapshots` configuration option with details like snapshot frequency and quantity before pruning, as well as S3-compatible connection information
  - [x] Daemon/worker thread to periodically run backups using VACUUM approach
- [x] S3 management to list/create snapshots (https://github.com/marcua/ayb/pull/317)
  - [x] Snapshot model with: primary key, database foreign key, creation date, snapshot ID (database hash/transaction ID, indexed), automatic/manual enum, [future not now: nullable note field for manual snapshot]
- [x] end-to-end tests using localstack as stand-in for S3 (https://github.com/marcua/ayb/pull/358)
- [x] Endpoints / client support for listing snapshots (#385)
- [x] Endpoints / client support for restoring from snapshots (#405)
- [x] Hash snapshot content to avoid uploading already-snapshotted database (#417 and #418)
- [x] Remove old snapshots in list beyond the retention quantity/period. (#418)
- [x] Replace gzip with zstd (#424)
- [x] Clean up: Initialize a HostedDb that has a SQLite / DuckDB implementation. Push query/backup logic into that. Consider doing this on an InstantiatedDatabase directly. (Turned into #426) 
- [x] Optimization based on modified time to reduce snapshots: get fs::metadata of each file in the dir, call `modified()` on result, sort the times so it's stable, take max or hash them together. (Turned into #427)
- [x] Final cleanups (#432)
  - [x] Clean up TODOs, especially in src/server/snapshots.rs
  - [x] Can we replace walkdir with std::fs::read_dir
- [x] Add a mutex to prevent overlapping/concurrent snapshot runs (#444)
- [ ] Write docs

## Alternatives we considered
* Knowing nothing,  [litestream](https://litestream.io/getting-started/) seemed like the most obvious option. The litestream documentation isn't clear on how many databases worth of replication it's designed for, but given that adding a database requires a configuration file update and a litestream restart, it was more likely designed around single digit to tens of databases. I roughly confirmed this in [a discussion in the libstream board](https://github.com/benbjohnson/litestream/discussions/561). The discussion raised the idea of waiting on a third party to open source one of their services for solving a similar problem, or calling Go from Rust, which we can consider in the future as those projects mature.
* In researching this work, the [libSQL](https://github.com/tursodatabase/libsql) project, and in particular 
[libsql-server/sqld](https://github.com/tursodatabase/libsql/tree/main/libsql-server) and its [bottomless replication support](https://github.com/tursodatabase/libsql/tree/main/bottomless) look promising. The bottomless documentation and commit history suggests its under quite a bit of heavy development. After all of my research, my intuition is that if the team ships a stable `bottomless` library, we'll be able to use it directly from `ayb` (both are in Rust) to solve the stream-the-write-ahead-log-to-and-recover-from-S3 problem.

Both projects, but in particular the second one, seem promising for the future, but weren't directly available for import at the time of writing, so for this v1 of backup/restore, we'll fall back to a [simpler solution like scheduled backups](https://litestream.io/alternatives/cron/).

**Comment by marcua** [comment]
Asked questions to understand litestream feasibility here: https://github.com/benbjohnson/litestream/discussions/561

**Comment by marcua** [comment]
Based on the response to https://github.com/benbjohnson/litestream/discussions/561, it looks like there isn't yet a programmatic way to add databases to replicate or restore, and that to use litestream for now, we'd have to update a growing .yml file and restart a service. I'd like to avoid that if possible, but keep watching the project.

I'm now thinking through two different paths
* Create a `Snapshot` data model that can store details on database snapshots (when we took it, what the hash of the database directory was, etc.) and then build scheduled backups as well as manual snapshot functionality around it. This would work well for most cases and be the simplest implementation path. The two downsides are: 1) Periodic snapshots have a larger window for data loss on the order of tens of minutes to an hour, and 2) For the first reason, implementing things like multi-node `ayb` support with failover will be difficult in the future.
* Look into [libsql-server/sqld](https://github.com/tursodatabase/libsql/tree/main/libsql-server) and its bottomless replication support (the docs have a broken link, but it claims to replicate the SQLite file and WAL to S3-compatible storage)

**Comment by marcua** [comment]
Some updates
* [.sha3sum --schema](https://www.sqlite.org/cli.html#sha3sum) is no longer the approach to beat --- it only works in the SQLite CLI. As a result, my current approach will be to have each DB in its own directory (SQLite spans multiple files), and use filesystem modified times / hashes of all files in the directory to determine changes.
* I realized I was overcomplicating the model by storing snapshot data in the ayb DB when the metadata already lived in S3. For now, I moved the snapshot model into a snapshot manager module, and S3 will be the source of truth (metadata can live in file headers). The only downside I can imagine is a performance one, which we should address if we encounter performance issues.
* This project has taken too long (it's a complicated one to revisit 2 hours a week, especially solo), and I plan to chunk it into tinier pieces to help with incremental progress/to reduce mental overhead
  * Refactors to directory structure that support snapshots and SQLite-in-directory
  * Refactors to allow dangerous/un-isolated requests to complete the backups (requires ATTACH)
  * Backups to the local filesystem  
  * S3 management to list/create snapshots
  * S3 management to retrieve snapshots
  * Refactor to run databases in isolation as mini-services rather than isolating on each request
  * Bootstrapping logic to check if recovery is necessary (database not available locally when you start the mini-service)
  * Optimizations based on modified time and hashes to reduce snapshots

**Comment by marcua** [comment]
Put/list appears to be working! For next time
- Clean up code (some S3 config and path syntax updates) and commit.
- Start on read/recovery path
- e2e testing
- Clean up running to-do list in snapshots.rs (it's dated now) and introduce any optimizations listed there

**Comment by marcua** [comment]
Code is cleaned up!
- Test force_path_style=true/false
- Address problem with `localstack` creating zombie python3 processes
- Make `localstack` hooks create a bucket or two for testing (or push this into integration tests)
- Start on read/recovery path
- e2e testing
- Clean up running to-do list in snapshots.rs (it's dated now) and introduce any optimizations listed there

**Comment by marcua** [comment]
Localstack crash issue: the `python3 ./bin/localstack status services --format=json` command is zombieing an increasing number of defunct `python3` processes

Will keep investigating

```
root       77626  0.1  0.0      0     0 ?        Z    23:45   0:00 [python3] <defunct>
root       77677  0.1  0.0      0     0 ?        Z    23:45   0:00 [python3] <defunct>
root       77721  0.1  0.0      0     0 ?        Z    23:45   0:00 [python3] <defunct>
root       77783  0.1  0.0      0     0 ?        Z    23:46   0:00 [python3] <defunct>
root       77809  0.1  0.0      0     0 ?        Z    23:46   0:00 [python3] <defunct>
root       77838  0.1  0.0      0     0 ?        Z    23:46   0:00 [python3] <defunct>
root       77862  0.1  0.0      0     0 ?        Z    23:46   0:00 [python3] <defunct>
root       77892  0.2  0.0      0     0 ?        Z    23:47   0:00 [python3] <defunct>
root       77918  0.2  0.0      0     0 ?        Z    23:47   0:00 [python3] <defunct>
root       77944  0.2  0.0      0     0 ?        Z    23:47   0:00 [python3] <defunct>
root       77970  0.2  0.0      0     0 ?        Z    23:47   0:00 [python3] <defunct>
root       77996  0.2  0.0      0     0 ?        Z    23:48   0:00 [python3] <defunct>
root       78023  0.2  0.0      0     0 ?        Z    23:48   0:00 [python3] <defunct>
root       78058  0.2  0.0      0     0 ?        Z    23:48   0:00 [python3] <defunct>
root       78087  0.3  0.0      0     0 ?        Z    23:49   0:00 [python3] <defunct>
root       78119  0.3  0.0      0     0 ?        Z    23:49   0:00 [python3] <defunct>
root       78145  0.3  0.0      0     0 ?        Z    23:49   0:00 [python3] <defunct>
root       78178  0.3  0.0      0     0 ?        Z    23:49   0:00 [python3] <defunct>
root       78207  0.4  0.0      0     0 ?        Z    23:50   0:00 [python3] <defunct>
root       78237  0.4  0.0      0     0 ?        Z    23:50   0:00 [python3] <defunct>
root       78265  0.4  0.0      0     0 ?        Z    23:50   0:00 [python3] <defunct>
root       78297  0.6  0.0      0     0 ?        Z    23:51   0:00 [python3] <defunct>
root       78330  0.7  0.0      0     0 ?        Z    23:51   0:00 [python3] <defunct>
root       78354  0.7  0.0      0     0 ?        Z    23:51   0:00 [python3] <defunct>
root       78377  1.0  0.0      0     0 ?        Z    23:51   0:00 [python3] <defunct>
root       78409  1.6  0.0      0     0 ?        Z    23:52   0:00 [python3] <defunct>
root       78439  2.6  0.0      0     0 ?        Z    23:52   0:00 [python3] <defunct>
root       78467  7.1  1.9  46100 39856 ?        S    23:52   0:00 python3 ./bin/localstack status services --format=json
```

**Comment by marcua** [comment]
Figured it out: a mere mortal would think they should start localstack with `localstack start`, but someone who knows The Truth will start it with `localstack start -d --network ls`. (The `-d` runs it in detached mode, less important, but the `--network ls` part fixed all of my Docker networking issues.)

- Make `localstack` hooks create a bucket or two for testing (or push this into integration tests)
- Test force_path_style=true/false
- Start on read/recovery path
- e2e testing
- Clean up running to-do list in snapshots.rs (it's dated now) and introduce any optimizations listed there

**Comment by marcua** [comment]
Now that localstack/force_path_style is in place, the remaining list is
- Read/recovery path
- e2e testing
- Clean up running to-do list in snapshots.rs (it's dated now) and introduce any optimizations listed there

**Comment by marcua** [comment]
Implementing e2e tests, leaving myself a to-do for the future
* Run localstack outside of tests (script should create a sqlite-tests and postgres-tests bucket)
* Configure ayb.tomls for sqlite/postgres to access their respective bucket in snapshot config
* Top of test_snapshots should delete bucket/create bucket
* Now that test can assume blank slate

**Comment by marcua** [comment]
Tests are starting to work --- we now delete the contents of the snapshot prefix we're going to be running tests on. Now it's time to validate the output (which contains timestamps, so requires regexes to match):
* Make utils::testing::list_snapshots match against a regex
* Finish basic "no snapshots exist, one exists after sleep, one still exists after sleep, update DB, two exist after sleep" 
* Start localstack in GH actions
* Clean up PR
* Next PR: restore from snapshot

**Comment by marcua** [comment]
Regex test of list_snapshots works!! Next up
- More gracefully handle snapshot loop trying to access DB directory before it exists
- Start localstack in GH actions
- Clean up PR
- Next PR: restore from snapshot

**Comment by marcua** [comment]
Wooohoo! Localstack now works in GH actions, and I cleaned up/standardized the e2e setup process and documented it. I submitted/reviewed/merged #385.

Next up: Functionality/endpoint for restoring from snapshot

**Comment by marcua** [comment]
* Restore works in terms of pulling the file down from S3
* But we want to be able to atomically swap the current database with the restored snapshot
* using std::fs:rename doesn't work, as the destination has to be an empty directory (it's not)
* it seems like the solution is a symlink
  * $db_path/current is a symlink to the current directory
  * $db_path/UUID is either a fresh database that was created, or a place where we move the restored snapshot
  * if current symlink exists, call std::fs::canonicalize on it to get its actual path, then create a new symlink with std::os::unix::fs::symlink / std::os::windows::fs::symlink_dir (will it overwrite the existing one like `ln -snf` does?), then remove the original (canonicalized) path.

**Comment by marcua** [comment]
#405 is shipped! Next up: hash to avoid unnecessary snapshots, and remove snapshots past the retention period!

**Comment by marcua** [comment]
Rough plan for hashing
* Write a method to, given a directory, walk every file in it in a deterministic order (i.e., sort by filename), hash the file's contents, and hash the hashes together.
* Compute a pre-hash
* List available snapshots and exit early if a snapshot matches the pre-hash
* Take a snapshot and compute a post-snapshot hash
* Exit early if a snapshot matches the post-hash

**Comment by marcua** [comment]
The more I think about it, prehashing is problematic
- if a database is actively being updated, the prehash might be of a different database entirely than the one in the snapshot
- the prehash prevents wasted work (snapshotting and hashing) in case the database never changed, but we can accomplish that more cheaply by keeping local filesystem modification times, only snapshotting locally and hashing the first time we run the snapshotting logic in a machine/process
- if we have there be only one hash (post-hash), we can change the identifier of the database to be its hash, which means we get the hash details in list_snapshots from S3 without having to issue subsequent head calls for metadata

So perhaps the path forward is
* Remove pre-/post-hash headers
* snapshot_id -> snapshot_hash, even in the name
* Write a method to, given a directory, walk every file in it in a deterministic order (i.e., sort by filename), hash the file's contents, and hash the hashes together. Blake3 seems to be [a cryptographic hash function that's faster than md5 and sha256](https://crates.io/crates/blake3).
* Take a snapshot and compute a post-snapshot hash
* Exit early if a snapshot in list_snapshots matches the hash
* In a future PR, if snapshotting is really expensive, add modification metadata to the snapshotting loop so that we don't needlessly snapshot/hash

**Comment by marcua** [comment]
#417 introduces hash-based identification of snapshots. The current implementation is naive in that it still uploads/overwrites a snapshot if one already exists with the key. From here:
* Take snapshot
* Compute hash
* Call list_snapshots
* If hash exists in list_snapshots, don't upload new one
* Update tests to ensure that modification date of that snapshot doesn't change even after waiting for another snapshot
* Since we have a snapshot list, if list_snapshots.len > max_snapshots, prune the oldest snapshot
* Update tests to ensure that we don't exceed the number of snapshots
* In a future PR, if snapshotting is really expensive, add modification metadata to the snapshotting loop so that we don't needlessly snapshot/hash

**Comment by marcua** [comment]
#417 is shipped, and #418 implements the hashing/pruning optimizations above! Just have to finished up some tests for #418, and then can move on to zstd/walkdir optimizations.

**Comment by marcua** [comment]
gzip -> zstd is done with #424!

next up is largely code cleanups!

**Comment by marcua** [comment]
Wow! The code cleanups are largely complete thanks to #432 and what I think remains before the next version is to add user-facing documentation. It's been so long!

**Comment by marcua** [comment]
With an asynchrony issue addressed in #444 and a draft of documentation availabe in #446, all that's left is reviewing the docs, making any updates, shipping the docs, and releasing the latest `ayb` to complete this issue!
